assets/src/asset.rs

```diff
@@ -0,0 +1,164 @@
+use std::error::Error;
+
+use fnv::FnvHashMap;
+use parking_lot::RwLock;
+
+use StoreId;
+
+/// One of the three core traits of this crate.
+///
+/// You want to implement this for every type of asset like
+///
+/// * `Mesh`
+/// * `Texture`
+/// * `Terrain`
+///
+/// and so on. Now, an asset may be available in different formats.
+/// That's why we have the `Data` associated type here. You can specify
+/// an intermediate format here, like the vertex data for a mesh or the samples
+/// for audio data.
+///
+/// This data is then generated by the `Format` trait.
+pub trait Asset
+    where Self: Sized
+{
+    type Context;
+    type Data;
+    type Error: Error;
+
+    /// A small keyword for which category this asset belongs to.
+    ///
+    /// ## Examples
+    ///
+    /// * `"mesh"` for `Mesh`
+    /// * `"data"` for `Level`
+    ///
+    /// The storage may use this information, to e.g. search the identically-named
+    /// subfolder.
+    fn category() -> &'static str;
+
+    /// Provides the conversion from the data format to the actual asset.
+    fn from_data(data: Self::Data, context: &Self::Context) -> Result<Self, Self::Error>;
+
+    /// Notifies about an asset load. This is can be used to cache the asset.
+    /// To return a cached asset, see the `cached` function.
+    fn asset_loaded(_context: &Self::Context, _spec: AssetSpec, _asset: &Self) {}
+
+    /// Returns `Some` cached value if possible, otherwise `None`.
+    ///
+    /// For a basic implementation of a cache, please take a look at the `Cache` type.
+    fn cached(_context: &Self::Context, _spec: &AssetSpec) -> Option<Self> {
+        None
+    }
+
+    /// Gives a hint that several assets may have been released recently.
+    ///
+    /// This is useful if your assets are reference counted, because you are
+    /// now able to remove unique assets from the cache, leaving the shared
+    /// ones there.
+    fn clear(_context: &Self::Context) {}
+
+    /// Request for clearing the whole cache.
+    fn clear_all(_context: &Self::Context) {}
+}
+
+/// A specifier for an asset, uniquely identifying it by
+///
+/// * the extension (the format it was provided in)
+/// * it's name
+/// * the storage it was loaded from
+#[derive(Clone, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct AssetSpec {
+    /// The extension of this asset
+    pub ext: &'static str,
+    /// The name of this asset.
+    pub name: String,
+    /// The storage id, uniquely identifying the storage it's been loaded from.
+    pub store: StoreId,
+}
+
+impl AssetSpec {
+    /// Creates a new asset specifier from the given parameters.
+    pub fn new(name: String, ext: &'static str, store: StoreId) -> Self {
```

Comment:
Would you be able to make this an `Into<String>` as well? I think I missed this on the last review.